Firestore/Source/API/FSTUserDataReader.mm

```diff
@@ -277,51 +318,70 @@ - (FieldValue)parsedQueryValue:(id)input allowArrays:(bool)allowArrays {
       if (context.array_element() && context.data_source() != UserDataSource::ArrayArgument) {
         ThrowInvalidArgument("Nested arrays are not supported");
       }
-      return [self parseArray:(NSArray *)input context:std::move(context)];
+      return absl::optional<google_firestore_v1_Value>([self parseArray:(NSArray *)input
+                                                                context:std::move(context)]);
     } else {
-      return [self parseScalarValue:input context:std::move(context)];
+      return absl::optional<google_firestore_v1_Value>([self parseScalarValue:input
+                                                                      context:std::move(context)]);
     }
   }
 }
 
-- (FieldValue)parseDictionary:(NSDictionary<NSString *, id> *)dict
-                      context:(ParseContext &&)context {
+- (google_firestore_v1_Value)parseDictionary:(NSDictionary<NSString *, id> *)dict
+                                     context:(ParseContext &&)context {
+  google_firestore_v1_Value result;
+  result.which_value_type = google_firestore_v1_Value_map_value_tag;
+
+  __block std::vector<google_firestore_v1_MapValue_FieldsEntry> fields;
+
   if (dict.count == 0) {
     const FieldPath *path = context.path();
     if (path && !path->empty()) {
       context.AddToFieldMask(*path);
     }
-    return ObjectValue::Empty().AsFieldValue();
   } else {
-    __block ObjectValue result = ObjectValue::Empty();
-
+    // Populate a vector of fields since we don't know the size of the final fields array.
     [dict enumerateKeysAndObjectsUsingBlock:^(NSString *key, id value, BOOL *) {
-      absl::optional<FieldValue> parsedValue =
+      absl::optional<google_firestore_v1_Value> parsedValue =
           [self parseData:value context:context.ChildContext(util::MakeString(key))];
       if (parsedValue) {
-        FieldPath path = FieldPath{util::MakeString(key)};
-        result = result.Set(path, *parsedValue);
+        google_firestore_v1_MapValue_FieldsEntry fieldsEntry;
+        fieldsEntry.key = nanopb::MakeBytesArray(util::MakeString(key));
+        fieldsEntry.value = *parsedValue;
+        fields.push_back(fieldsEntry);
       }
     }];
+  }
 
-    return result;
+  result.map_value.fields_count = static_cast<pb_size_t>(fields.size());
+  result.map_value.fields =
+      nanopb::MakeArray<google_firestore_v1_MapValue_FieldsEntry>(result.map_value.fields_count);
+
+  for (size_t i = 0; i < fields.size(); ++i) {
+    result.map_value.fields[i] = fields[i];
   }
+
+  return result;
 }
 
-- (FieldValue)parseArray:(NSArray<id> *)array context:(ParseContext &&)context {
-  __block FieldValue::Array result;
-  result.reserve(array.count);
+- (google_firestore_v1_Value)parseArray:(NSArray<id> *)array context:(ParseContext &&)context {
+  __block google_firestore_v1_Value result;
```

Comment:
Same note re. using `result{}`. The idea is similar to go/totw/182.

Reply:
Done